Bogus "Function is missing a type annotation for one or more arguments" with attrs

When checking this file with mypy (0.641 or git master) and `--disallow-incomplete-defs`:

```python
import attr

@attr.s
class A:
    a = attr.ib()
```

I get an unexpected:

```
at.py:3: error: Function is missing a type annotation for one or more arguments
```

(line 3 being the `@attr.s` decorator). If I'm interpreting [what `--disallow-incomplete-defs` does](https://mypy.readthedocs.io/en/latest/command_line.html#untyped-definitions-and-calls) correctly, it should only warn if there's a partial type annotation (as opposed to none at all, which `--disallow-untyped-defs` would warn about).

Also, the error message is confusing - I don't define a function at all there. With `--show-error-context`, it says:

```
at.py: note: In member "__init__" of class "A":
```

FWIW, I (finally) started using mypy with a rather big project ([qutebrowser](https://qutebrowser.org/)) and I'm quite happy with it so far - the approach of gradual typing and the awesome documentation make it very easy to get a benefit without having to annotate the entire codebase at once.

> > The first one is not really an option, because the following code is unannotated and we don't want it to be magically marked as fully annotated
> 
> Perhaps this is okay if the plugin tells the user that the class attribute is unannotated? It seems that would make understanding the error a lot easier to understand.

That's already done with `--disallow-untyped-defs`:

```
at.py: note: In member "__init__" of class "A":
at.py:3: error: Function is missing a type annotation for one or more arguments
at.py: note: In class "A":
at.py:5: error: Need type annotation for 'a'
```

but with `--disallow-incomplete-defs` and no annotations at all, I'd expect no warning/error.

--- a/mypy/plugins/attrs.py
+++ b/mypy/plugins/attrs.py
@@ -639,11 +639,21 @@ def _make_frozen(ctx: 'mypy.plugin.ClassDefContext', attributes: List[Attribute]
 def _add_init(ctx: 'mypy.plugin.ClassDefContext', attributes: List[Attribute],
               adder: 'MethodAdder') -> None:
     """Generate an __init__ method for the attributes and add it to the class."""
-    adder.add_method(
-        '__init__',
-        [attribute.argument(ctx) for attribute in attributes if attribute.init],
-        NoneType()
-    )
+    args = [attribute.argument(ctx) for attribute in attributes if attribute.init]
+    if all(
+        # We use getattr rather than instance checks because the variable.type
+        # might be wrapped into a Union or some other type, but even non-Any
+        # types reliably track the fact that the argument was not annotated.
+        getattr(arg.variable.type, "type_of_any", None) == TypeOfAny.unannotated
+        for arg in args
+    ):
+        # This workaround makes --disallow-incomplete-defs usable with attrs,
+        # but is definitely suboptimal as a long-term solution.
+        # See https://github.com/python/mypy/issues/5954 for discussion.
+        for a in args:
+            a.variable.type = AnyType(TypeOfAny.implementation_artifact)
+            a.type_annotation = AnyType(TypeOfAny.implementation_artifact)
+    adder.add_method('__init__', args, NoneType())
 
 
 class MethodAdder:

--- a/test-data/unit/check-attr.test
+++ b/test-data/unit/check-attr.test
@@ -75,7 +75,7 @@ A(1, [2], '3', 4, 5)  # E: Too many arguments for "A"
 # flags: --disallow-untyped-defs
 import attr
 @attr.s
-class A:  # E: Function is missing a type annotation for one or more arguments
+class A:
     a = attr.ib()  # E: Need type annotation for 'a'
     _b = attr.ib()  # E: Need type annotation for '_b'
     c = attr.ib(18)  # E: Need type annotation for 'c'
@@ -1121,7 +1121,7 @@ reveal_type(B)  # N: Revealed type is 'def (x: __main__.C) -> __main__.B'
 import attr
 
 @attr.s
-class B:  # E: Function is missing a type annotation for one or more arguments
+class B:
     x = attr.ib()  # E: Need type annotation for 'x'
 
 reveal_type(B)  # N: Revealed type is 'def (x: Any) -> __main__.B'

--- a/test-data/unit/check-flags.test
+++ b/test-data/unit/check-flags.test
@@ -1029,6 +1029,31 @@ def f(i: int, s):
 main:3: error: Function is missing a return type annotation
 main:3: error: Function is missing a type annotation for one or more arguments
 
+[case testDisallowIncompleteDefsAttrsNoAnnotations]
+# flags: --disallow-incomplete-defs
+import attr
+
+@attr.s()
+class Unannotated:
+    foo = attr.ib()
+
+[case testDisallowIncompleteDefsAttrsWithAnnotations]
+# flags: --disallow-incomplete-defs
+import attr
+
+@attr.s()
+class Annotated:
+    bar: int = attr.ib()
+
+[case testDisallowIncompleteDefsAttrsPartialAnnotations]
+# flags: --disallow-incomplete-defs
+import attr
+
+@attr.s()
+class PartiallyAnnotated:  # E: Function is missing a type annotation for one or more arguments
+    bar: int = attr.ib()
+    baz = attr.ib()
+
 [case testAlwaysTrueAlwaysFalseFlags]
 # flags: --always-true=YOLO --always-true=YOLO1 --always-false=BLAH1 --always-false BLAH --ignore-missing-imports
 from somewhere import YOLO, BLAH

